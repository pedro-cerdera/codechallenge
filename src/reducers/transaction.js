const initialState = {
  ammount: 0,
  transactions: [],
  isLoading: false,
  error: null,
};

export const Types = {
  TRANSCATION_ADD: '@transaction/ADD',
  TRANSCATION_SET_LOADING: '@transaction/SET_LOADING',
  TRANSCATION_SET_ERROR: '@transaction/SET_ERROR',
  TRANSCATION_RESET: '@transaction/RESET',
};

export const Actions = {};

export const TransactionReducer = (state = initialState, action) => {
  switch (action.type) {
    case Types.TRANSCATION_SET_LOADING:
      return {
        ...state,
        isLoading: action.payload,
      };
    case Types.TRANSCATION_ADD:
      return {
        ...state,
        transactions: [...state.transactions, action.payload],
        ammount: state.ammount + action.payload?.value,
      };
    case Types.TRANSCATION_SET_ERROR:
      return {
        ...state,
        error: action.payload,
      };
    case Types.TRANSCATION_RESET:
      return initialState;

    default:
      return {...state};
  }
};
