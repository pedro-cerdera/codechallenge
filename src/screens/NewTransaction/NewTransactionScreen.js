
import React from 'react';
import { Text, View } from 'react-native';

const NewTransactionScreen = () => (
  <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
    <Text>New Transaction Screen</Text>
  </View>
);


export default NewTransactionScreen;