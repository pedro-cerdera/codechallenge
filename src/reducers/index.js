import {TransactionReducer, Actions as TranscationActions} from './transaction';

export default {TransactionReducer};

export const Actions = {Transaction: TranscationActions};
