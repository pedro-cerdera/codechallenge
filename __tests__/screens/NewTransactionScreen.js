import React from 'react';
import {fireEvent, render} from 'react-native-testing-library';
import {Types} from '../../src/reducers/transaction';
import NewTransactionScreen from '../../src/screens/NewTransaction/NewTransactionScreen';

jest.mock('../../src/assets/images/logo.png', () => ({
  uri: 'logo.png',
}));
jest.mock(
  '../../src/components/OperationsList/OperationsList',
  () => 'OperationsList',
);

const mockDispacth = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispacth,
}));

const mockNavigation = jest.fn();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    goBack: mockNavigation,
  }),
}));

describe('Testing HomeScreen', () => {
  it('renders correctly', () => {
    const tree = render(<NewTransactionScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('new request', () => {
    const {getByTestId} = render(<NewTransactionScreen />);
    fireEvent.changeText(getByTestId('value'), '20');
    fireEvent.changeText(getByTestId('description'), 'Test');
    fireEvent.press(getByTestId('button'));

    expect(mockDispacth).toHaveBeenCalledWith({
      type: Types.TRANSCATION_ADD,
      payload: {
        value: -20,
        description: 'Test',
      },
    });
    expect(mockDispacth).toHaveBeenCalledTimes(1);
    expect(mockNavigation).toHaveBeenCalledTimes(1);
  });
});
