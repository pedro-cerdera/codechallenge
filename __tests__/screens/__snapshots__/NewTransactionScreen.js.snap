// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing HomeScreen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#1C2124",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
  testID="new-request-screen"
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 20,
        },
        Object {
          "alignItems": "center",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": "logo.png",
        }
      }
    />
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flex": 1,
            "maxHeight": 350,
            "width": "100%",
          },
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "borderRadius": 5,
              "elevation": 1,
              "maxWidth": 500,
              "shadowColor": "gray",
              "shadowOffset": Object {
                "height": 3,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 3,
              "width": "100%",
            },
            Object {
              "flex": 1,
              "justifyContent": "space-between",
              "padding": 20,
            },
          ]
        }
      >
        <RCTScrollView
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "display": "flex",
                  "marginTop": 0,
                }
              }
            >
              <Text>
                Valor
              </Text>
              <TextInput
                allowFontScaling={true}
                keyboardType="numeric"
                masked={true}
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="R$ 100,00"
                placeholderTextColor="#999999"
                rejectResponderTermination={true}
                style={
                  Array [
                    Object {
                      "backgroundColor": "white",
                      "borderColor": "black",
                      "borderRadius": 25,
                      "borderStyle": "solid",
                      "borderWidth": 1,
                      "height": 50,
                      "justifyContent": "center",
                      "marginBottom": 0,
                      "marginLeft": 0,
                      "marginRight": 0,
                      "marginTop": 0,
                      "maxHeight": 200,
                      "paddingBottom": 15,
                      "paddingLeft": 19,
                      "paddingRight": 19,
                      "paddingTop": 15,
                    },
                  ]
                }
                testID="value"
                title="Valor"
                topSpacing={false}
                type="money"
                underlineColorAndroid="transparent"
                value="R$0,00"
                visible={true}
              />
            </View>
            <View
              style={
                Object {
                  "display": "flex",
                  "marginTop": 15,
                }
              }
            >
              <Text>
                Descrição
              </Text>
              <TextInput
                allowFontScaling={true}
                keyboardType="default"
                masked={false}
                multiline={true}
                numberOfLines={2}
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="Ex.: Transferência bancária"
                placeholderTextColor="#999999"
                rejectResponderTermination={true}
                style={
                  Array [
                    Object {
                      "backgroundColor": "white",
                      "borderColor": "black",
                      "borderRadius": 25,
                      "borderStyle": "solid",
                      "borderWidth": 1,
                      "justifyContent": "center",
                      "marginBottom": 0,
                      "marginLeft": 0,
                      "marginRight": 0,
                      "marginTop": 0,
                      "maxHeight": 200,
                      "minHeight": 50,
                      "paddingBottom": 15,
                      "paddingLeft": 19,
                      "paddingRight": 19,
                      "paddingTop": 15,
                    },
                  ]
                }
                testID="description"
                title="Descrição"
                topSpacing={true}
                underlineColorAndroid="transparent"
                value=""
                visible={true}
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginTop": 15,
                }
              }
            >
              <Text>
                Saque
              </Text>
              <RCTSwitch
                accessibilityRole="switch"
                onChange={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "height": 31,
                    "width": 51,
                  }
                }
                testID="switch"
                value={false}
              />
              <Text>
                Depósito
              </Text>
            </View>
          </View>
        </RCTScrollView>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#1C2124",
              "borderRadius": 25,
              "opacity": 1,
              "padding": 15,
              "width": "100%",
            }
          }
          testID="button"
        >
          <Text
            style={
              Object {
                "color": "white",
              }
            }
          >
            Fazer Transação
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 30,
        }
      }
    />
  </View>
</RCTSafeAreaView>
`;
